File: backend/api/notification_initializer.py
"""
Notification Initializer Service
Automatically creates relevant notifications for users
"""

from .mongodb_service import NotificationService
from datetime import datetime, timedelta
import logging

logger = logging.getLogger(__name__)

class NotificationInitializer:
    """Service for initializing user notifications"""
    
    def __init__(self):
        self.notification_service = NotificationService()
    
    def _to_object_id(self, user_id: str):
        """Convert string to ObjectId"""
        from bson import ObjectId
        if isinstance(user_id, str):
            return ObjectId(user_id)
        return user_id
    
    def initialize_user_notifications(self, user_id: str) -> bool:
        """Initialize notifications for a new or existing user"""
        try:
            # Check if user already has a notification bundle
            existing_bundle = self.notification_service.db.notifications.find_one({
                "user_id": self.notification_service._to_object_id(user_id),
                "type": "bundle"
            })
            
            if existing_bundle:
                logger.info(f"User {user_id} already has notification bundle, skipping initialization")
                return True
            
            # Create all notification messages
            all_messages = []
            all_messages.extend(self._get_welcome_messages())
            all_messages.extend(self._get_tip_messages())
            all_messages.extend(self._get_reminder_messages())
            
            # Create single notification bundle
            bundle_id = self.notification_service.create_user_notification_bundle(user_id, all_messages)
            
            if bundle_id:
                logger.info(f"Successfully created notification bundle {bundle_id} for user {user_id}")
                return True
            else:
                logger.error(f"Failed to create notification bundle for user {user_id}")
                return False
            
        except Exception as e:
            logger.error(f"Error initializing notifications for user {user_id}: {e}")
            return False
    
    def _get_welcome_messages(self):
        """Get welcome messages for new users"""
        return [
            {
                "type": "welcome",
                "title": "🎉 Welcome to Financability!",
                "message": "Welcome to your personal finance management platform. Start by setting up your budget and tracking your expenses.",
                "priority": "high",
                "is_read": False
            },
            {
                "type": "getting_started",
                "title": "📊 Getting Started Guide",
                "message": "Check out the Dashboard to see your financial overview, or visit Debt Planning to create your debt payoff strategy.",
                "priority": "medium",
                "is_read": False
            }
        ]
    
    def _get_tip_messages(self):
        """Get helpful tip messages"""
        return [
            {
                "type": "tip",
                "title": "💡 Pro Tip: Emergency Fund",
                "message": "Aim to build an emergency fund that covers 3-6 months of expenses for financial security.",
                "priority": "low",
                "is_read": False
            },
            {
                "type": "tip",
                "title": "💡 Pro Tip: Debt Snowball vs Avalanche",
                "message": "Snowball method: pay smallest debts first. Avalanche method: pay highest interest first. Choose what motivates you!",
                "priority": "low",
                "is_read": False
            }
        ]
    
    def _get_reminder_messages(self):
        """Get reminder messages"""
        return [
            {
                "type": "reminder",
                "title": "🎯 Set Financial Goals",
                "message": "Set specific, measurable financial goals to stay motivated on your journey to financial freedom.",
                "priority": "medium",
                "is_read": False
            }
        ]
    
    def create_budget_alert_notification(self, user_id: str, category: str, spent: float, limit: float):
        """Create a budget alert notification"""
        over_budget = spent > limit
        percentage = (spent / limit) * 100 if limit > 0 else 0
        
        if over_budget:
            title = f"⚠️ Budget Alert: {category}"
            message = f"You've exceeded your {category} budget by ${spent - limit:.2f} ({percentage:.1f}% over limit)."
            priority = "high"
        else:
            title = f"💰 Budget Update: {category}"
            message = f"You've spent ${spent:.2f} of ${limit:.2f} in {category} ({percentage:.1f}% used)."
            priority = "medium"
        
        notification_data = {
            "type": "budget_alert",
            "title": title,
            "message": message,
            "priority": priority,
            "data": {
                "category": category,
                "spent": spent,
                "limit": limit,
                "percentage": percentage,
                "over_budget": over_budget
            }
        }
        
        return self.notification_service.create_notification(user_id, notification_data)
    
    def create_debt_milestone_notification(self, user_id: str, debt_name: str, milestone: str):
        """Create a debt milestone notification"""
        notification_data = {
            "type": "debt_milestone",
            "title": f"🎉 Debt Milestone: {debt_name}",
            "message": f"Congratulations! You've reached a milestone: {milestone}",
            "priority": "high",
            "data": {
                "debt_name": debt_name,
                "milestone": milestone
            }
        }
        
        return self.notification_service.create_notification(user_id, notification_data)
    
    def create_savings_goal_notification(self, user_id: str, goal_name: str, current_amount: float, target_amount: float):
        """Create a savings goal notification"""
        percentage = (current_amount / target_amount) * 100 if target_amount > 0 else 0
        
        if percentage >= 100:
            title = f"🎯 Goal Achieved: {goal_name}"
            message = f"Congratulations! You've reached your {goal_name} goal of ${target_amount:.2f}!"
            priority = "high"
        elif percentage >= 75:
            title = f"🏁 Almost There: {goal_name}"
            message = f"You're {percentage:.1f}% towards your {goal_name} goal. Keep it up!"
            priority = "medium"
        else:
            title = f"💪 Progress Update: {goal_name}"
            message = f"You're {percentage:.1f}% towards your {goal_name} goal (${current_amount:.2f} of ${target_amount:.2f})."
            priority = "low"
        
        notification_data = {
            "type": "savings_goal",
            "title": title,
            "message": message,
            "priority": priority,
            "data": {
                "goal_name": goal_name,
                "current_amount": current_amount,
                "target_amount": target_amount,
                "percentage": percentage
            }
        }
        
        return self.notification_service.create_notification(user_id, notification_data)

File: financability-mobile/src/services/notificationInitializer.ts
import notificationService from './notificationService';
import { CreateNotificationData } from './notificationService';

class NotificationInitializer {
  /**
   * Initialize notifications for a new user
   */
  async initializeUserNotifications(): Promise<boolean> {
    try {
      console.log('🔔 Initializing user notifications...');
      const response = await notificationService.initializeNotifications();
      console.log('🔔 Notifications initialized:', response);
      return true;
    } catch (error) {
      console.error('Failed to initialize notifications:', error);
      return false;
    }
  }

  /**
   * Create a budget alert notification
   */
  async createBudgetAlert(category: string, spent: number, limit: number): Promise<string | null> {
    try {
      const alertData = {
        category,
        spent,
        limit,
      };
      
      const notificationId = await notificationService.createBudgetAlert(alertData);
      console.log('🔔 Budget alert created:', notificationId);
      return notificationId;
    } catch (error) {
      console.error('Failed to create budget alert:', error);
      return null;
    }
  }

  /**
   * Create a debt reminder notification
   */
  async createDebtReminder(debtName: string, dueDate: string, amount: number): Promise<string | null> {
    try {
      const reminderData = {
        debt_name: debtName,
        due_date: dueDate,
        amount,
      };
      
      const notificationId = await notificationService.createDebtReminder(reminderData);
      console.log('🔔 Debt reminder created:', notificationId);
      return notificationId;
    } catch (error) {
      console.error('Failed to create debt reminder:', error);
      return null;
    }
  }

  /**
   * Create a savings milestone notification
   */
  async createSavingsMilestone(goal: string, progress: number, target: number): Promise<string | null> {
    try {
      const milestoneData = {
        goal,
        progress,
        target,
      };
      
      const notificationId = await notificationService.createSavingsMilestone(milestoneData);
      console.log('🔔 Savings milestone created:', notificationId);
      return notificationId;
    } catch (error) {
      console.error('Failed to create savings milestone:', error);
      return null;
    }
  }

  /**
   * Create a custom notification
   */
  async createCustomNotification(
    type: string,
    title: string,
    message: string,
    priority: 'low' | 'medium' | 'high' = 'medium',
    data?: Record<string, any>
  ): Promise<string | null> {
    try {
      const notificationData: CreateNotificationData = {
        type,
        title,
        message,
        priority,
        data,
      };
      
      const notificationId = await notificationService.createNotification(notificationData);
      console.log('🔔 Custom notification created:', notificationId);
      return notificationId;
    } catch (error) {
      console.error('Failed to create custom notification:', error);
      return null;
    }
  }

  /**
   * Create welcome notifications for new users
   */
  async createWelcomeNotifications(): Promise<void> {
    try {
      const welcomeNotifications = [
        {
          type: 'welcome',
          title: '🎉 Welcome to Financability!',
          message: 'Welcome to your personal finance management platform. Start by setting up your budget and tracking your expenses.',
          priority: 'high' as const,
        },
        {
          type: 'getting_started',
          title: '📊 Getting Started Guide',
          message: 'Check out the Dashboard to see your financial overview, or visit Debt Planning to create your debt payoff strategy.',
          priority: 'medium' as const,
        },
        {
          type: 'tip',
          title: '💡 Pro Tip: Emergency Fund',
          message: 'Aim to build an emergency fund that covers 3-6 months of expenses for financial security.',
          priority: 'low' as const,
        },
        {
          type: 'tip',
          title: '💡 Pro Tip: Debt Snowball vs Avalanche',
          message: 'Snowball method: pay smallest debts first. Avalanche method: pay highest interest first. Choose what motivates you!',
          priority: 'low' as const,
        },
      ];

      for (const notification of welcomeNotifications) {
        await this.createCustomNotification(
          notification.type,
          notification.title,
          notification.message,
          notification.priority
        );
      }
    } catch (error) {
      console.error('Failed to create welcome notifications:', error);
    }
  }

  /**
   * Create financial milestone notifications
   */
  async createFinancialMilestoneNotifications(): Promise<void> {
    try {
      const milestoneNotifications = [
        {
          type: 'milestone',
          title: '🏆 Financial Milestone Achieved!',
          message: 'Congratulations! You\'ve reached a significant financial milestone. Keep up the great work!',
          priority: 'high' as const,
        },
        {
          type: 'reminder',
          title: '⏰ Monthly Budget Review',
          message: 'It\'s time to review your monthly budget and make adjustments for the upcoming month.',
          priority: 'medium' as const,
        },
        {
          type: 'reminder',
          title: '💳 Debt Payment Reminder',
          message: 'Don\'t forget to make your debt payments on time to avoid late fees and maintain your credit score.',
          priority: 'medium' as const,
        },
      ];

      for (const notification of milestoneNotifications) {
        await this.createCustomNotification(
          notification.type,
          notification.title,
          notification.message,
          notification.priority
        );
      }
    } catch (error) {
      console.error('Failed to create financial milestone notifications:', error);
    }
  }

  /**
   * Create educational tip notifications
   */
  async createEducationalTipNotifications(): Promise<void> {
    try {
      const tipNotifications = [
        {
          type: 'tip',
          title: '📚 Financial Education: Compound Interest',
          message: 'Compound interest is the eighth wonder of the world. Start investing early to take advantage of this powerful force.',
          priority: 'low' as const,
        },
        {
          type: 'tip',
          title: '📚 Financial Education: 50/30/20 Rule',
          message: 'Allocate 50% for needs, 30% for wants, and 20% for savings and debt repayment.',
          priority: 'low' as const,
        },
        {
          type: 'tip',
          title: '📚 Financial Education: Credit Score',
          message: 'Your credit score affects loan rates and insurance premiums. Pay bills on time and keep credit utilization low.',
          priority: 'low' as const,
        },
      ];

      for (const notification of tipNotifications) {
        await this.createCustomNotification(
          notification.type,
          notification.title,
          notification.message,
          notification.priority
        );
      }
    } catch (error) {
      console.error('Failed to create educational tip notifications:', error);
    }
  }
}

export default new NotificationInitializer();

